README.md
[![Tests workflow](https://img.shields.io/github/workflow/status/mahozad/theme-switch/CI?label=Tests&logo=github)](https://github.com/mahozad/theme-switch/actions/workflows/ci.yml)
[![Codecov](https://img.shields.io/codecov/c/gh/mahozad/theme-switch?label=Coverage&logo=codecov&logoColor=%23FF56C0&token=C4P4I1TQTF)](https://codecov.io/gh/mahozad/theme-switch)
[![Minified size](https://img.shields.io/github/size/mahozad/theme-switch/dist/theme-switch.min.js?label=Minified%20size)](https://unpkg.com/@mahozad/theme-switch)
[![Published on webcomponents.org](https://img.shields.io/badge/webcomponents.org-published-9351dc.svg)](https://www.webcomponents.org/element/@mahozad/theme-switch)

<div align="center">

![Animated icon](https://raw.githubusercontent.com/mahozad/theme-switch/main/icon.svg)

</div>

# HTML light/dark/system theme switch button

This is an animated [custom HTML element](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements)
which toggles between light theme, dark theme, and automatic theme (OS theme).
It adds a custom attribute (`data-theme`) to the `html` element of your page.
You can style your page the way you like based on the value of that attribute.

See the [demo page](https://mahozad.ir/theme-switch/) and scroll below for an [example](#styling-a-page-based-on-the-theme).

## Installing and using the library

  - ### Regular HTML pages
    ```html
    <script src="https://unpkg.com/@mahozad/theme-switch"></script>
    ```
  - ### Node.js and npm
    ```shell
    npm install --save @mahozad/theme-switch
    ```
    ```html
    <script src="node_modules/@mahozad/theme-switch/dist/theme-switch.min.js"></script>
    ```

  - ### Angular and more
    For Angular framework and more details about the above installation methods see the [wiki](https://github.com/mahozad/theme-switch/wiki).

After installing the library, use the element like this (make sure to include the end tag):

```html
<theme-switch></theme-switch>
```

## Styling the switch element

A custom element is no different from the built-in elements of HTML.  
Use and style it however you want just like you would use and style a regular element:

```css
theme-switch {
  width: 64px;
  padding: 8px;
  background: #888;
  
  /* There is a special property called --theme-switch-icon-color
   * which you can set, to change the color of the icon in the switch */
  --theme-switch-icon-color: #aabbcc;
}
```

## Styling a page based on the theme

In your CSS stylesheet, specify your desired styles for light and dark themes.
One way is to define [custom CSS properties](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties)
for your colors, sizes, etc. and redefine them (if needed) with new values for the other theme:

```css
/* These are applied for the default (light) theme */
/* (or when the toggle is auto, and the OS theme is light) */
html {
  --my-page-background-color: #fff;
  --my-text-color: red;
}

/* These are applied for the dark theme */
/* (or when the toggle is auto, and the OS theme is dark) */
/* If a property has the same value for both light and dark themes, no need to redeclare it here */
html[data-theme="dark"] {
  --my-page-background-color: #112233;
}

body /* or any selector you want */ {
  background: var(--my-page-background-color);
  color: var(--my-text-color);
}
```

## Misc

<details>

<summary>Click here to expand</summary>

The switch element fires (triggers) a custom event called `themeToggle` every time it is toggled (clicked).
You can listen and react to it if you want:

```javascript
document.addEventListener("themeToggle", event => {
  console.log(`Old theme: ${ event.detail.oldState }`);
  console.log(`New theme: ${ event.detail.newState }`);
  // More operations...
});
```

---

This widget was inspired by [this YouTube video](https://youtu.be/kZiS1QStIWc)
and [this library](https://github.com/GoogleChromeLabs/dark-mode-toggle).

---

See [this article](https://css-tricks.com/web-components-are-easier-than-you-think/)
which is about creating HTML custom elements.

See the icon for switching themes (located in the top right corner) on
[Google Fonts site](https://fonts.google.com/icons).  
Also see [this site](https://rastikerdar.github.io/vazirmatn).

See [this article](https://css-tricks.com/a-complete-guide-to-dark-mode-on-the-web)
for implementing dark/light theme on sites.

See [this post](https://stackoverflow.com/q/56300132/8583692) for how to override
dark/light theme for a site.

See [this comprehensive GitHub repo](https://github.com/mateusortiz/webcomponents-the-right-way) about custom elements.

---

### Similar libraries
  - [Dark Mode Toggle](https://github.com/H0rn0chse/dark-mode-toggle)
  - [\<dark-mode-toggle>](https://github.com/GoogleChromeLabs/dark-mode-toggle)
  - [\<color-scheme-button>](https://github.com/CICCIOSGAMINO/color-scheme-button)
  - [\<theme-toggle>](https://github.com/mothepro/theme-toggle)
  - [Binary theme switcher component](https://github.com/diegosanchezp/theme-switcher-component)

---

TODO:
  - Try to add the library to [rufus site](https://github.com/pbatard/rufus-web)
  - Try to add the library to [jest site](https://github.com/facebook/jest) (probably its `docs/` directory. see [this PR](https://github.com/facebook/jest/pull/11021))
  - Try to add the library to [MDN site](https://developer.mozilla.org/en-US/)
  - Try to add the library to [docusaurus](https://github.com/facebook/docusaurus)
  - Try to add the library to [dokka](https://github.com/Kotlin/dokka)
  - Try to add the library to [mkdocs-material](https://github.com/squidfunk/mkdocs-material)
  - See [chrome auto dark feature for android](https://developer.chrome.com/blog/new-in-chrome-98/#autodark-opt-out)

</details>


CHANGELOG.md
# History of notable changes introduced in each version

## v1.5.2 (2022-03-31)
#### Updates
  - Revert the change to the main field in package.json introduced in the previous version ([`e5c36003`](https://github.com/mahozad/theme-switch/commit/e5c36003))

[All commits since version 1.5.1](https://github.com/mahozad/theme-switch/compare/v1.5.1...v1.5.2)

## v1.5.1 (2022-03-30)
#### Updates
  - Change the library main script to the minified version (in hope of fixing the bundlephobia badge) ([`0de6d2df`](https://github.com/mahozad/theme-switch/commit/0de6d2df))

[All commits since version 1.5.0](https://github.com/mahozad/theme-switch/compare/v1.5.0...v1.5.1)

## v1.5.0 (2022-03-30)
#### New features
  - Reflect updates to switches in one page to all other open pages ([`b5586f00`](https://github.com/mahozad/theme-switch/commit/b5586f00))

[All commits since version 1.4.0](https://github.com/mahozad/theme-switch/compare/v1.4.0...v1.5.0)

## v1.4.0 (2022-03-14)
#### Updates
  - Update the animation of the "auto" icon ([`ae30ad8a`](https://github.com/mahozad/theme-switch/commit/ae30ad8a))

[All commits since version 1.3.0](https://github.com/mahozad/theme-switch/compare/v1.3.0...v1.4.0)

## v1.3.0 (2022-03-13)
#### Other
  - Migrate from Babel to [Rollup](https://rollupjs.org/) ([`c6f0f209`](https://github.com/mahozad/theme-switch/commit/c6f0f209), [`7f02c996`](https://github.com/mahozad/theme-switch/commit/7f02c996), and others).
    This results in a smaller minified script.
    Note that from this version on, the files to be used are in *dist/* directory of the library.

[All commits since version 1.2.1](https://github.com/mahozad/theme-switch/compare/v1.2.1...v1.3.0)

## v1.2.1 (2022-03-11)
#### Other
  - Add some code refactorings ([`0a2aab86`](https://github.com/mahozad/theme-switch/commit/0a2aab86), [`3d5573a2`](https://github.com/mahozad/theme-switch/commit/3d5573a2), [`7fc2b960`](https://github.com/mahozad/theme-switch/commit/7fc2b960))

[All commits since version 1.2.0](https://github.com/mahozad/theme-switch/compare/v1.2.0...v1.2.1)

## v1.2.0 (2022-03-08)
#### New features
  - Add previous and new theme name in the custom event ([`a34af81a`](https://github.com/mahozad/theme-switch/commit/a34af81a))
#### Other
  - Allow invalid strings as the value of theme key in storage ([`453bc60b`](https://github.com/mahozad/theme-switch/commit/453bc60b))
  - Add some minor code improvements ([`6d8bc9c9`](https://github.com/mahozad/theme-switch/commit/6d8bc9c9), [`074b33fa`](https://github.com/mahozad/theme-switch/commit/074b33fa), [`8c3082b1`](https://github.com/mahozad/theme-switch/commit/8c3082b1), [`104c2ef2`](https://github.com/mahozad/theme-switch/commit/104c2ef2))

[All commits since version 1.1.0](https://github.com/mahozad/theme-switch/compare/v1.1.0...v1.2.0)

## v1.1.0 (2022-03-07)
#### New features
  - Add support for multiple instances of the element in a page ([`aa435053`](https://github.com/mahozad/theme-switch/commit/aa435053));
    if there are multiple instances, they are all updated/animated in synchrony when one is toggled ([`8811803e`](https://github.com/mahozad/theme-switch/commit/8811803e))
  - A new custom event called `themeToggle` is triggered whenever the element is toggled. It can be listened to if needed ([`8811803e`](https://github.com/mahozad/theme-switch/commit/8811803e))
#### Bug fixes
  - Make the element able to be hidden (if/when `display: none` is applied to the element) ([`32a0f5d4`](https://github.com/mahozad/theme-switch/commit/32a0f5d4))

[All commits since version 1.0.0](https://github.com/mahozad/theme-switch/compare/v1.0.0...v1.1.0)

## v1.0.0 (2022-03-06)
#### Bug fixes
  - Fix the bug with sun rays color not being affected by the `--theme-switch-icon-color` CSS property ([`c90d7874`](https://github.com/mahozad/theme-switch/commit/c90d7874))

[All commits since version 1.0.0-rc](https://github.com/mahozad/theme-switch/compare/v1.0.0-rc...v1.0.0)

## v1.0.0-rc (2022-03-06)
#### Removals
  - Remove the unnecessary Node.js exports ([`cd296e55`](https://github.com/mahozad/theme-switch/commit/cd296e55))

[All commits since version 0.9.2](https://github.com/mahozad/theme-switch/compare/v0.9.2...v1.0.0-rc)

## v0.9.2 (2022-03-04)
#### Bug fixes
  - Wrap the whole code (minified version) in an IIFE to avoid exposing variables and functions in global scope ([`a659c312`](https://github.com/mahozad/theme-switch/commit/a659c312))

[All commits since version 0.9.1](https://github.com/mahozad/theme-switch/compare/v0.9.1...v0.9.2)

## v0.9.1 (2022-03-01)
This version had no changes to the code.

[All commits since version 0.9.0](https://github.com/mahozad/theme-switch/compare/v0.9.0...v0.9.1)

## v0.9.0 (2021-03-01)
This is the first release of the library.


[comment]: <> (NOTE: Be aware that modifying the format of this file might impact the script that makes the body of GitHub releases)


# Template:
## vx.y.z (yyyy-mm-dd)
#### New features
  - new feature 1
#### Updates
  - change 1
#### Bug fixes
  - bug fix 1
#### Deprecations
  - deprecation 1
#### Removals
  - removal 1
#### Other
  - other 1


theme-switch.test.js
const {JSDOM} = require("jsdom");
const puppeteer = require("puppeteer-core");
const fileSystem = require('fs');
// See https://stackoverflow.com/a/48952855/8583692
const {configureToMatchImageSnapshot: configureSnapshots} = require("jest-image-snapshot");
const toMatchReferenceSnapshot = configureSnapshots({
    customSnapshotsDir: "snapshots/",
    customDiffDir: "snapshot-diffs/",
});
const snapshotFileName = "temp-snapshot-for-test.png";
/**
 * See https://stackoverflow.com/q/7163061/8583692
 * Add a local.json file with the content like this:
 * {
 *   "chromiumPath": "C:\\Program Files\\Google\\Chrome\\Application\\chrome.exe"
 * }
 * The file is intentionally ignored in VCS.
 */
const {chromiumPath} = require("./local.json");

expect.extend({ toMatchReferenceSnapshot });

// TODO: Also test for the following cases:
//  - No <theme-switch> element
//  - Duplicate <theme-switch> elements
// See https://github.com/chaijs/type-detect/issues/98
// See https://stackoverflow.com/a/51702674/8583692
// This HTML is used to create window, localstorage, etc. and is used for unit tests.
// Screenshot tests use their own HTML file.
const dom = new JSDOM(`
    <!DOCTYPE html>
    <html lang="en">
    <head>
      <meta charset="UTF-8">
      <title>Test page</title>
    </head>
    <body>
      <theme-switch></theme-switch>
    </body>
    </html> 
`,
    {url: "http://localhost"}
);

global.window = dom.window;
global.document = dom.window.document;
global.HTMLElement = dom.window.HTMLElement;
global.localStorage = dom.window.localStorage;
setSystemThemeTo("light");

const main = require("./theme-switch");
// Could have instead exported the functions in theme-switch.js
const mainInternals = {
    toggleTheme: main.__get__("toggleTheme"),
    getSystemTheme: main.__get__("getSystemTheme"),
    getUserThemeSelection: main.__get__("getUserThemeSelection")
};

test(`When system theme is light, getSystemTheme should return "light"`, () => {
    expect(mainInternals.getSystemTheme()).toBe("light");
});

test(`When system theme is dark, getSystemTheme should return "dark"`, () => {
    setSystemThemeTo("dark");
    expect(mainInternals.getSystemTheme()).toBe("dark");
});

test(`For first-time users, getUserThemeSelection should return the default theme`, () => {
    localStorage.clear();
    expect(mainInternals.getUserThemeSelection()).toBe("light");
});

test(`For first-time users, getUserThemeSelection should return the default theme irrespective of the system theme`, () => {
    localStorage.clear();
    setSystemThemeTo("dark");
    expect(mainInternals.getUserThemeSelection()).toBe("light");
});

test(`getUserThemeSelection should return "light" when user had previously selected "light"`, () => {
    localStorage.setItem("theme", "light");
    expect(mainInternals.getUserThemeSelection()).toBe("light");
});

test(`getUserThemeSelection should return "dark" when user had previously selected "dark"`, () => {
    localStorage.setItem("theme", "dark");
    expect(mainInternals.getUserThemeSelection()).toBe("dark");
});

test(`getUserThemeSelection should return "auto" when user had previously selected "auto"`, () => {
    localStorage.setItem("theme", "auto");
    expect(mainInternals.getUserThemeSelection()).toBe("auto");
});

/**
 * NOTE: jsdom doesn't seem to support `beginElement()` function on SVG `animate` element.
 *  See https://github.com/jsdom/jsdom/issues/3344
 *  and https://github.com/jsdom/jsdom/issues/2647
 *  and https://stackoverflow.com/q/44173754/8583692
 *  We had ho mock the implementation of some functions to prevent error.
 *
 * We cannot mock an internal function of a module with Jest.
 * Instead, we used babel rewire plugin
 * (or rewire-test-env-only plugin so the minified file is not bloated
 * see https://www.npmjs.com/package/babel-plugin-rewire-test-env-only).
 * See https://stackoverflow.com/q/51269431/8583692
 * and https://stackoverflow.com/q/51900413/8583692
 * and https://github.com/speedskater/babel-plugin-rewire
 * and https://github.com/jhnns/rewire/issues/136#issuecomment-380829197
 * and https://www.npmtrends.com/babel-plugin-rewire-vs-mock-require-vs-proxyquire-vs-rewire
 *
 * If we wanted to mock a regular function, we could have used any of these approaches:
 *
 * ```javascript
 * jest.mock("theme-switch");
 * jest.spyOn(main, "animateThemeButtonIconToDark").mockImplementation(() => jest.fn());
 * jest.mock("theme-switch", () => ({
 *     ...jest.requireActual("theme-switch"),
 *     animateThemeButtonIconToDark: jest.fn()
 * }));
 * main.animateThemeButtonIconToDark.mockImplementation(() => {})
 * main.animateThemeButtonIconToDark = jest.fn(() => {});
 * ```
 */
test(`When user theme is light, toggleTheme should update the theme to dark`, () => {
    main.__Rewire__("animateThemeButtonIconToDark", () => {});
    localStorage.setItem("theme", "light");
    mainInternals.toggleTheme();
    expect(mainInternals.getUserThemeSelection()).toBe("dark");
});

test(`When user theme is dark, toggleTheme should update the theme to auto`, () => {
    main.__Rewire__("animateThemeButtonIconToAuto", () => {});
    localStorage.setItem("theme", "dark");
    mainInternals.toggleTheme();
    expect(mainInternals.getUserThemeSelection()).toBe("auto");
});

test(`When user theme is auto, toggleTheme should update the theme to light`, () => {
    main.__Rewire__("animateThemeButtonIconToLight", () => {});
    localStorage.setItem("theme", "auto");
    mainInternals.toggleTheme();
    expect(mainInternals.getUserThemeSelection()).toBe("light");
});

describe("Screenshot tests", () => {
    // Increase the timeout of executing all the test suit from
    // the default 5000 to a greater value to run fine on CI
    // Also, set the timeout per test for each individual test
    // See https://github.com/facebook/jest/issues/5055
    jest.setTimeout(100_000);

    // See https://stackoverflow.com/a/53299842/8583692
    // Consider Selenium as an alternative to Puppeteer.
    // Selenium doesn't seem to support screenshot testing feature.
    // See https://stackoverflow.com/q/22938045/8583692
    test(`When user stored theme is light, the icon should be sun`, async () => {
        await takeScreenshot(() => {localStorage.setItem("theme", "light");});
        const snapshotTakenNow = fileSystem.readFileSync(snapshotFileName);
        expect(snapshotTakenNow).toMatchReferenceSnapshot();
    }, 100_000);

    test(`When user stored theme is dark, the icon should be moon`, async () => {
        await takeScreenshot(() => {localStorage.setItem("theme", "dark");});
        const snapshotTakenNow = fileSystem.readFileSync(snapshotFileName);
        expect(snapshotTakenNow).toMatchReferenceSnapshot();
    }, 100_000);

    test(`When user stored theme is auto, the icon should be auto`, async () => {
        await takeScreenshot(() => {localStorage.setItem("theme", "auto");});
        const snapshotTakenNow = fileSystem.readFileSync(snapshotFileName);
        expect(snapshotTakenNow).toMatchReferenceSnapshot();
    }, 100_000);

    test(`When user stored theme is light, clicking the switch should change the icon to moon`, async () => {
        await takeScreenshot(
            () => {localStorage.setItem("theme", "light");},
            element => {element.click();}
        );
        const snapshotTakenNow = fileSystem.readFileSync(snapshotFileName);
        expect(snapshotTakenNow).toMatchReferenceSnapshot();
    }, 100_000);

    test(`When user stored theme is dark, clicking the switch should change the icon to auto`, async () => {
        await takeScreenshot(
            () => {localStorage.setItem("theme", "dark");},
            element => {element.click();}
        );
        const snapshotTakenNow = fileSystem.readFileSync(snapshotFileName);
        expect(snapshotTakenNow).toMatchReferenceSnapshot();
    }, 100_000);

    test(`When user stored theme is auto, clicking the switch should change the icon to light`, async () => {
        await takeScreenshot(
            () => {localStorage.setItem("theme", "auto");},
            element => {element.click();}
        );
        const snapshotTakenNow = fileSystem.readFileSync(snapshotFileName);
        expect(snapshotTakenNow).toMatchReferenceSnapshot();
    }, 100_000);

    test(`When user specifies a custom color for switch icon, the colors should be applied`, async () => {
        await takeScreenshot(
            () => {localStorage.setItem("theme", "light");},
            element => {
                // See https://stackoverflow.com/a/64487791/8583692
                element.evaluate((el) => {
                    el.style.setProperty("--theme-switch-icon-color", "#ffe36e");
                });
            }
        );
        const snapshotTakenNow = fileSystem.readFileSync(snapshotFileName);
        expect(snapshotTakenNow).toMatchReferenceSnapshot();
    }, 100_000);

    test(`When user specifies a custom color for switch icon in a CSS rule with low specificity (like html{}), the colors should be applied`, async () => {
        await takeScreenshot(
            () => {localStorage.setItem("theme", "light");},
            () => {},
            "test2.html"
        );
        const snapshotTakenNow = fileSystem.readFileSync(snapshotFileName);
        expect(snapshotTakenNow).toMatchReferenceSnapshot();
    }, 100_000);

    // Make the element able to be hidden (with CSS property display: none)
    // See https://developers.google.com/web/fundamentals/web-components/best-practices
    // and https://stackoverflow.com/q/47144187/8583692
    test(`When user specifies hidden attribute on the element, element should be hidden`, async () => {
        const action = async () => {
            await takeScreenshot(() => {}, () => {}, "test3.html");
        };
        await expect(action).rejects.toThrowError("Node is either not visible or not an HTMLElement");
    }, 100_000);

    afterAll(() => {fileSystem.rmSync(snapshotFileName);});
});

async function takeScreenshot(init, action = () => {}, pageHTML = "test1.html") {
    const browser = await puppeteer.launch({
            headless: true, // If false, opens the browser UI
            // channel: "chrome", // this overrides executablePath
            // Download the required version from https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html?prefix=Win_x64
            // OR C:\\your_workspace\\node_modules\\puppeteer\\.local-chromium\\win64-(version)\\chrome-win\\chrome.exe
            executablePath: chromiumPath
        }
    );

    const page = await browser.newPage();
    // See https://stackoverflow.com/a/66530593/8583692
    await page.evaluateOnNewDocument(init);
    // page.setContent("<DOCTYPE html><html>...")
    await page.goto(`file://${__dirname}\\${pageHTML}`);

    const element = await page.$("theme-switch");
    await action(element);
    // Wait for the action or element animation to finish
    await page.waitForTimeout(1000);

    try {
        await element.screenshot({path: snapshotFileName});
    } finally {
        // NOTE: This call should be in finally block so if taking screenshot
        //  threw any error then the browser is closed no matter what to avoid
        //  the tests not being finished (running forever)
        await browser.close();
    }
}

// console.log(`\u001B[32m✔️\u001B[39m Tests passed`);

// See https://stackoverflow.com/a/53449595/8583692
// and https://stackoverflow.com/a/57180950/8583692
function setSystemThemeTo(mode) {
    const isDark = mode === "dark";
    Object.defineProperty(window, "matchMedia", {
        writable: true,
        value: jest.fn().mockImplementation(query => ({
            matches: isDark,
            media: query,
            onchange: null,
            addListener: jest.fn(), // Deprecated
            removeListener: jest.fn(), // Deprecated
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        })),
    });
}


src/main.js
/*
* NOTE: Do not use this script as an ES6 module.
*  ES6 modules are deferred and we don't want that because
*  we want the user previous theme selection to be applied
*  as soon as possible (before the page is rendered).
* */

/*
* There are two types of modules mostly used in JavaScript.
* One is created by Node.js and is used inside the Node environment
* and has been available for a long time. It is called CommonJS.
* Another is the standard native JavaScript modules introduced in ES6.
*
* The Node variant (CommonJS) uses `module.exports` (or simply `exports`) and
* `require()` to export and import scripts, functions, variables, etc.
* Browsers do not know about `exports` or `require` functions and throw error
* because they are objects and functions created just in Node environment and set globally.
* If you want to use this type of module in browsers, you should bundle the files
* (merge all of them into a single JS file which eliminates the need for exports and require)
* with tools like babel, webpack, rollup, etc.
*
* ES6 modules use `export` and `import` keywords for the same purpose.
*
* Example Node modules:
*
* my-calculator.js
* const PI = 3.14;
* function calculate() {}
* modules.exports.calculate = calculate;
* modules.exports.PI = PI;
*
* main.js
* const calculator = require("my-calculator");
* const perimeter = 2 * calculator.PI;
* const result = calculator.calculate();
*
* Example ES6 modules:
*
* my-calculator.js
* export const PI = 3.14;
* export function calculate() {}
*
* main.js
* import { calculate, PI } from "my-calculator.js";
* const perimeter = 2 * PI;
* const result = calculate();
*
* See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules
* and https://stackoverflow.com/a/9901097/8583692
* */

/*
* Minify the script either through command line with babel:
* - babel main.js --source-maps --out-file result.min.js
* or with babel-minify (also has an alias called minify) which is useful if
* you don't already use babel (as a preset) or want to run minification standalone.
* Note that it does not take into account babel.config.json settings.
* - babel-minify (or minify) main.js --mangle --no-comments --out-file result.min.js`
* Or automate it with IntelliJ file watcher
* - babel
*   + program: $ProjectFileDir$\node_modules\.bin\babel
*   + arguments: $FilePathRelativeToProjectRoot$ --out-file $FileNameWithoutExtension$.min.js --source-maps
*   Note that setting "sourceMaps": "true" in babel.config.json does not work because of
*   this bug: https://github.com/babel/babel/issues/5261 ("sourceMaps": "inline" works, however)
* - UglifyJS
* - YUI compressor (seems to be deprecated and removed in newer versions of IntelliJ)
*
* Babel preset-env inserts a semicolon at the start of the minified file.
* See why: https://stackoverflow.com/q/1873983/8583692
* */

// TODO: extract Jest configuration to a jest.config.js file
// TODO: Add an attribute so the user can define key name stored in localstorage
// TODO: Make the component customizable by adding custom attributes:
//  See https://medium.com/technofunnel/creating-passing-data-to-html-custom-elements-using-attributes-bfd9aa759fd4

// FIXME: If the switch is toggled too quickly, the switches in other open tabs may not update consistently.
//  Steps to reproduce the bug:
//  - Open two tabs
//  - Set the theme of both tabs to "auto"
//  - In one of the pages, click the switch three times in rapid succession so its new state is "auto" again
//    The other tab should also show auto icon but it shows light icon.
//  To resolve this, instead of listening for localstorage events, maybe we can pull the
//  localstorage state for changes every 100 ms and update the switch if necessary.

/*
* NOTE: To avoid name collisions if another script declares variables or functions with the same name
*  as ours (i.e. defining them in the global scope) and browsers complaining about identifiers
*  being redeclared, we wrap all our code in a closure or IIFE (sort of creating a namespace for it).
*  ES6 now supports block scope as well (simply wrapping the whole code in {}).
*  I am now using the babel-plugin-iife-wrap plugin to wrap the whole result (minified)
*  code in an IIFE.
* For examples, see these libraries:
*   - https://github.com/highlightjs/highlight.js/blob/main/src/highlight.js
*   - https://github.com/jashkenas/underscore/blob/master/underscore.js
* See
*   - https://www.w3schools.com/js/js_scope.asp
*   - https://github.com/jhnns/rewire/issues/136#issuecomment-380829197
*   - https://stackoverflow.com/a/32750216/8583692
*   - https://stackoverflow.com/q/8228281/8583692
*   - https://stackoverflow.com/q/881515/8583692
*   - https://stackoverflow.com/q/39388777/8583692
*   - https://stackoverflow.com/a/47207686/8583692
* We could also do something like these libraries:
*   - https://github.com/juliangarnier/anime/blob/master/build.js
*   - https://github.com/mrdoob/three.js/
*   - https://github.com/moment/moment
*   - https://github.com/floating-ui/floating-ui
* */

const ELEMENT_NAME = "theme-switch";
const ICON_SIZE = 24 /* px */;
const ICON_COLOR = "#000";
const THEME_KEY = "theme";
const THEME_AUTO = "auto";
const THEME_DARK = "dark";
const THEME_LIGHT = "light";
const THEME_VALUES = [THEME_AUTO, THEME_DARK, THEME_LIGHT];
const THEME_DEFAULT = THEME_LIGHT;
const THEME_ATTRIBUTE = "data-theme";
const COLOR_SCHEME_DARK = "(prefers-color-scheme: dark)";
const CUSTOM_EVENT_NAME = "themeToggle";
// circleRadius, raysOpacity, eclipseCenterX, letterOffset
const ICON_INITIAL_STATE_FOR_AUTO = [10, 0, 33, 0];
const ICON_INITIAL_STATE_FOR_DARK = [10, 0, 20, 1];
const ICON_INITIAL_STATE_FOR_LIGHT = [5, 1, 33, 1];

class ThemeSwitchElement extends HTMLElement {
    shadowRoot;
    static counter = 0; // See https://stackoverflow.com/a/43116254/8583692
    identifier = ThemeSwitchElement.counter++;

    constructor() {
        super();
        // See https://stackoverflow.com/q/2305654/8583692
        this.shadowRoot = this.attachShadow({ mode: "open" });
        this.shadowRoot.innerHTML = generateIcon(...getInitialStateForIcon());
        // Add the click listener to the top-most parent (the custom element itself)
        // so the padding etc. on the element be also clickable
        this.shadowRoot.host.addEventListener("click", this.onClick);
        // If another theme switch in page toggled, update my icon too
        document.addEventListener(CUSTOM_EVENT_NAME, event => {
            if (event.detail.originId !== this.identifier) {
                this.adaptToTheme();
            }
        });
        // If a theme switch in another page toggled, update my state too
        // See https://developer.mozilla.org/en-US/docs/Web/API/Window/storage_event
        window.addEventListener("storage", event => {
            if (event.key === THEME_KEY) {
                this.adaptToTheme();
                updateTheme();
            }
        });
        // Create some CSS to apply to the shadow DOM
        // See https://css-tricks.com/styling-a-web-component/
        const style = document.createElement("style");
        style.textContent = generateStyle();
        this.shadowRoot.append(style);
    }

    onClick() {
        const oldTheme = getUserThemeSelection();
        this.toggleTheme(oldTheme);
        const newTheme = getUserThemeSelection();
        const event = this.createEvent(oldTheme, newTheme);
        this.dispatchEvent(event);
    }

    // See https://stackoverflow.com/a/53804106/8583692
    createEvent(oldTheme, newTheme) {
        return new CustomEvent(CUSTOM_EVENT_NAME, {
            detail: {
                originId: this.identifier,
                oldState: oldTheme,
                newState: newTheme
            },
            bubbles: true,
            composed: true,
            cancelable: false
        });
    }

    // See https://stackoverflow.com/q/48316611
    toggleTheme(currentTheme) {
        if (currentTheme === THEME_AUTO) {
            localStorage.setItem(THEME_KEY, THEME_LIGHT);
            this.animateThemeButtonIconToLight();
        } else if (currentTheme === THEME_DARK) {
            localStorage.setItem(THEME_KEY, THEME_AUTO);
            this.animateThemeButtonIconToAuto();
        } else /* if (theme === THEME_LIGHT) */ {
            localStorage.setItem(THEME_KEY, THEME_DARK);
            this.animateThemeButtonIconToDark();
        }
        updateTheme();
    }

    adaptToTheme() {
        const theme = getUserThemeSelection();
        if (theme === THEME_AUTO) {
            this.animateThemeButtonIconToAuto();
        } else if (theme === THEME_DARK) {
            this.animateThemeButtonIconToDark();
        } else /* if (theme === THEME_LIGHT) */ {
            this.animateThemeButtonIconToLight();
        }
    }

    animateThemeButtonIconToLight() {
        this.shadowRoot.getElementById("letter-anim-hide").beginElement();
        this.shadowRoot.getElementById("core-anim-shrink").beginElement();
        this.shadowRoot.getElementById("rays-anim-rotate").beginElement();
        this.shadowRoot.getElementById("rays-anim-show").beginElement();
    }

    animateThemeButtonIconToAuto() {
        this.shadowRoot.getElementById("eclipse-anim-go").beginElement();
        this.shadowRoot.getElementById("letter-anim-show").beginElement();
    }

    animateThemeButtonIconToDark() {
        this.shadowRoot.getElementById("rays-anim-hide").beginElement();
        this.shadowRoot.getElementById("core-anim-enlarge").beginElement();
        this.shadowRoot.getElementById("eclipse-anim-come").beginElement();
    }
}

function generateIcon(circleRadius, raysOpacity, eclipseCenterX, letterOffset) {
    return `ICON_TEMPLATE`;
}

function generateStyle() {
    return `STYLES_TEMPLATE`;
}

updateTheme();
window.customElements.define(ELEMENT_NAME, ThemeSwitchElement);
window
    .matchMedia(COLOR_SCHEME_DARK)
    .addEventListener("change", updateTheme);

function updateTheme() {
    let theme = getUserThemeSelection();
    if (theme === THEME_AUTO) theme = getSystemTheme();
    document.documentElement.setAttribute(THEME_ATTRIBUTE, theme);
}

function getUserThemeSelection() {
    const userSelection = localStorage.getItem(THEME_KEY);
    return THEME_VALUES.includes(userSelection) ? userSelection : THEME_DEFAULT;
}

function getSystemTheme() {
    const isDark = window.matchMedia(COLOR_SCHEME_DARK).matches;
    return isDark ? THEME_DARK : THEME_LIGHT;
}

function getInitialStateForIcon() {
    const theme = getUserThemeSelection();
    if (theme === THEME_AUTO) {
        return ICON_INITIAL_STATE_FOR_AUTO;
    } else if (theme === THEME_DARK) {
        return ICON_INITIAL_STATE_FOR_DARK;
    } else /* if (theme === THEME_LIGHT) */ {
        return ICON_INITIAL_STATE_FOR_LIGHT;
    }
}

// Export for tests run by npm (no longer needed; kept for future reference)
// See https://stackoverflow.com/q/63752210/8583692
// and https://stackoverflow.com/a/54680602/8583692
// and https://stackoverflow.com/q/43042889/8583692
// and https://stackoverflow.com/a/1984728/8583692
// if (typeof module !== "undefined") {
//     module.exports = {
//         updateTheme,
//         toggleTheme,
//         getSystemTheme,
//         getInitialStateForIcon,
//         animateThemeButtonIconToAuto,
//         animateThemeButtonIconToDark,
//         animateThemeButtonIconToLight
//     };
// }
